client/cli/src/commands/runcmd.rs

```diff
@@ -579,6 +551,25 @@ fn parse_cors(s: &str) -> std::result::Result<Cors, Box<dyn std::error::Error>>
 	})
 }
 
+#[derive(Debug, derive_more::Display)]
+pub struct InvalidRpcMethod(
+	#[display(fmt="Invalid RPC methods value: {}")]
+	String
+);
+impl std::error::Error for InvalidRpcMethod {}
+
+fn parse_rpc_methods(value: &str) -> std::result::Result<RpcMethods, InvalidRpcMethod> {
```

Comment:
Is custom parser necessary? I thought that structopt does support enums out of the box, no?

Reply:
I think it's only if you use the macro arg_enum

Reply:
Once I remove the custom parser and `parse(try_from_str = ...)` it's grumbling about unimplemented `FromStr`, so I'll just do directly

Reply:
I tried that and remembered why I did things this way - `RpcMethods` are implemented in `sc_service` so I can't implement `FromStr for RpcMethods` directly in `runcmd.rs`. However, I'd rather keep this impl here since it's only used here for parsing.

Do you mind if we leave it like that or should I add a `FromStr` impl near the `RpcMethods` definition?

Reply:
Fine to leave it like that. I think in some other cases we actually have two instances of the enum and we convert from one to another at some point, but that seems even more complicated.

Reply:
I just started wondering one thing. With custom parser how will the user know what options are available? I.e. Does the help display anything? If possible options are not generated in current case I'd rather go for a separate enum with `FromStr` implementation.

Reply:
> I think in some other cases we actually have two instances of the enum and we convert from one to another at some point, but that seems even more complicated.

Yes that's how it's done at the moment (I don't have opinion on this):

https://github.com/paritytech/substrate/blob/97ad80269db3acde1e2349ee96aefa5f517cf385/client/cli/src/arg_enums.rs#L22-L57

> I just started wondering one thing. With custom parser how will the user know what options are available? I.e. Does the help display anything?

You need to define `possible_values` in `#[structopt(...)]`.

Note that with arg_enum it's more or less automatically done for you.

Reply:
Eventually, I went with the two enums with a dedicated `arg_enum!`-wrapped one in arg_enums.rs. Thanks!